src/games/strategy/ui/SwingWorkerCompletionWaiter.java

```diff
@@ -0,0 +1,56 @@
+package games.strategy.ui;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import java.awt.Window;
+import java.beans.PropertyChangeEvent;
+import java.beans.PropertyChangeListener;
+
+import javax.swing.SwingWorker;
+
+/**
+ * Manages the display of a window during the execution of a {@code SwingWorker}.
+ *
+ * <p>
+ * Use this class when it is necessary to display progress during a potentially long-running operation. If the <i>Event
+ * Dispatch Thread</i> must be blocked until the operation is complete (i.e. the user should not be allowed to perform
+ * any other actions in the UI), it is recommended you use a modal dialog.
+ * </p>
+ *
+ * <p>
+ * For example:
+ * </p>
+ *
+ * <pre>
+ * JDialog dialog = new JDialog(owner, true);
+ * swingWorker.addPropertyChangeListener(
+ *     new SwingWorkerCompletionWaiter(dialog));
+ * swingWorker.execute();
+ * // the dialog will be visible until the SwingWorker is done
+ * dialog.setVisible(true);
+ * </pre>
+ */
+public final class SwingWorkerCompletionWaiter implements PropertyChangeListener {
+  private final Window window;
+
+  /**
+   * Initializes a new instance of the {@code SwingWorkerCompletionWaiter} class.
+   *
+   * @param window The window to display while the operation is in progress; must not be {@code null}.
+   */
+  public SwingWorkerCompletionWaiter(final Window window) {
+    this.window = checkNotNull(window);
+  }
+
+  @Override
+  public void propertyChange(final PropertyChangeEvent event) {
+    if ("state".equals(event.getPropertyName())) {
```

Comment:
"state" is the name of the property `SwingWorker` changes when the state of the background task changes.  Ideally, it would be defined as a constant on `SwingWorker`, but it's unfortunately not.  I'll extract a package-private constant in `SwingWorkerCompletionWaiter` so

- it avoids a magic number, and
- it can be reused by the corresponding test (which also uses the same literal string).